### PR TITLE
Allow the user to add a salt to the hash so he can invalidate the cac…

### PR DESCRIPTION
**Pull request by eyal-resonai**
…he when needed

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -7,6 +7,9 @@ before_install:
   - sudo add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable"
   - sudo apt-get update
   - sudo apt-get -y -o Dpkg::Options::="--force-confnew" install docker-ce
+  # the docker user and password are defined in
+  # https://app.travis-ci.com/github/resonai/ybt/settings
+  - echo "$DOCKER_PASSWORD" | docker login -u "$DOCKER_USERNAME" --password-stdin
```
Comment (itamaro) [inline .travis.yml:12]: why is this login needed?

File: yabt/target_utils.py [hunk 2]
```diff
@@ -235,6 +235,10 @@ def compute_target_json(self, build_context, prop_blacklist, deps_hashes):
             flavor=build_context.conf.flavor,  # TODO: any other conf args?
             # yabt_version=__version__,  # TODO: is this needed?
         )
+        if build_context.conf.get('cache_salt', None) is not None:
+            json_dict['cache_salt'] = build_context.conf.get('cache_salt',
+                                                             None),
+
```
Comment (itamaro) [inline yabt/target_utils.py:241]: nit - delete blank line

**Review by itamaro — APPROVED** [review]
code looks good, but can you explain the use-case and motivation to do this?

File: .travis.yml (hunk 1 above)
Comment (eyal-resonai) [reply .travis.yml:12]: docker-hub is getting more and more aggressive about users creating account and choose a plan so they started limiting the non-logged-in pulls and we are getting:
```
Error response from daemon: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
```
see https://app.travis-ci.com/github/resonai/ybt/jobs/554916402

File: yabt/target_utils.py (hunk 2 above)
Comment (eyal-resonai) [reply yabt/target_utils.py]: done

**Comment by eyal-resonai** [comment]
> code looks good, but can you explain the use-case and motivation to do this?

hi :)

The previous PR allowed the user of ybt to control the root directory of the proto builder generated code output (in order to overcome the problem of 3rd part module named proto: https://github.com/googleapis/proto-plus-python/issues/79).

Now that we can change that in our project's YSetting we are facing the problem that the cache hash has not changed but we need to invalidate all the proto targets cache.

We thought about adding salt for each builder and control it from there.
While it is a good idea to have and we do want to have it, having a user control salt is a must so that the user can easily invalidate cache when she changes things in YSettings that needs cache invalidation.

**Pull request merged**